salt/cloud/clouds/profitbricks.py

```diff
@@ -0,0 +1,836 @@
+            'when trying to run the initial deployment: \n{1}'.format(
+                vm_['name'], exc
+            ),
+            exc_info_on_loglevel=logging.DEBUG
+        )
+        return False
+
+    create_network_interfaces(conn, datacenter_id, data['id'], vm_)
+    set_public_lan(conn, vm_)
+
+    def __query_node_data(vm_, data):
+        running = False
+        try:
+            data = show_instance(vm_['name'], 'action')
+            if not data:
+                return False
+            log.debug(
+                'Loaded node data for {0}:\nname: {1}\nstate: {2}'.format(
+                    vm_['name'],
+                    pprint.pformat(data['name']),
+                    data['vmState']
+                )
+            )
+        except Exception as err:
+            log.error(
+                'Failed to get nodes list: {0}'.format(
+                    err
+                ),
+                # Show the trackback if the debug logging level is enabled
+                exc_info_on_loglevel=logging.DEBUG
+            )
+            # Trigger a failure in the wait for IP function
+            return False
+
+        running = data['vmState'] == 'RUNNING'
+        if not running:
+            # Still not running, trigger another iteration
+            return
+
+        if ssh_interface(vm_) == 'private_lan' and data['private_ips']:
+            vm_['ssh_host'] = data['private_ips'][0]
+
+        if ssh_interface(vm_) != 'private_lan' and data['public_ips']:
+            vm_['ssh_host'] = data['public_ips'][0]
+
+        return data
+
+    try:
+        data = salt.utils.cloud.wait_for_ip(
+            __query_node_data,
+            update_args=(vm_, data),
+            timeout=config.get_cloud_config_value(
+                'wait_for_ip_timeout', vm_, __opts__, default=10 * 60),
+            interval=config.get_cloud_config_value(
+                'wait_for_ip_interval', vm_, __opts__, default=10),
+        )
+    except (SaltCloudExecutionTimeout, SaltCloudExecutionFailure) as exc:
+        try:
+            # It might be already up, let's destroy it!
+            destroy(vm_['name'])
+        except SaltCloudSystemExit:
+            pass
+        finally:
+            raise SaltCloudSystemExit(str(exc.message))
+
+    log.debug('VM is now running')
+    log.info('Created Cloud VM {0[name]!r}'.format(vm_))
+    log.debug(
+        '{0[name]!r} VM creation details:\n{1}'.format(
+            vm_, pprint.pformat(data)
+        )
+    )
+
+    salt.utils.cloud.fire_event(
+        'event',
+        'created instance',
+        'salt/cloud/{0}/created'.format(vm_['name']),
+        {
+            'name': vm_['name'],
+            'profile': vm_['profile'],
+            'provider': vm_['driver'],
+        },
+        transport=__opts__['transport']
+    )
+
+    if 'ssh_host' in vm_:
+        vm_['key_filename'] = get_key_filename(vm_)
+        ret = salt.utils.cloud.bootstrap(vm_, __opts__)
```

Comment:
The ProfitBricks driver was originally created without SSH key support as the ProfitBricks platform did not provide public key injection at the time. So originally nodes could be provisioned, but not bootstrapped. SSH key support was then added recently and I included bootstrapping of the node. I thought I would fallback to allowing the node to be provisioned without bootstrapping if there were no keys available.

Is that behavior OK as long as I provide a warning? Or is it better to always require bootstrapping and exit if no keys are present?